components/suggest/src/db.rs

```diff
@@ -123,24 +123,45 @@ impl<'a> SuggestDao<'a> {
         Self { conn, scope }
     }
 
-    /// Fetches suggestions that match the given keyword from the database.
-    pub fn fetch_by_keyword(&self, keyword: &str) -> Result<Vec<Suggestion>> {
-        let (keyword_prefix, keyword_suffix) = split_keyword(keyword);
-        Ok(self.conn.query_rows_and_then_cached(
-            "SELECT s.id, k.rank, s.title, s.url, s.provider, NULL as confidence, NULL as keyword_suffix
-             FROM suggestions s
-             JOIN keywords k ON k.suggestion_id = s.id
-             WHERE k.keyword = :keyword
-             UNION ALL
-             SELECT s.id, k.rank, s.title, s.url, s.provider, k.confidence, k.keyword_suffix
-             FROM suggestions s
-             JOIN pocket_keywords k ON k.suggestion_id = s.id
-             WHERE k.keyword_prefix = :keyword_prefix",
-            named_params! {
-                ":keyword": keyword,
-                ":keyword_prefix": keyword_prefix,
-            },
-            |row| -> Result<Option<Suggestion>> {
+    /// Fetches suggestions that match the given query from the database.
+    pub fn fetch_suggestions(&self, query: &SuggestionQuery) -> Result<Vec<Suggestion>> {
+        let (keyword_prefix, keyword_suffix) = split_keyword(&query.keyword);
+
+        let (mut statement, params) = if query.providers.contains(&SuggestionProvider::Pocket) {
```

Comment:
Pocket keywords use a separate query, so, if we aren't matching Pocket suggestions, we can skip querying `pocket_keywords` entirely.

That also means we end up with a dynamically-sized list of `params`—two for Pocket; one for non-Pocket—so we have to use a `Vec<(&str, &dyn ToSql)>` to hold them.